common/app/model/Cached.scala

```diff
@@ -41,33 +42,32 @@ object Cached extends implicits.Dates {
     if (cacheableStatusCodes.contains(result.header.status)) cacheHeaders(cacheSeconds, result) else result
   }
 
-  case class StringResult(result: Result, body: Option[String])
-  object StringResult {
-    def apply(result: Result, body: String) = new StringResult(result, Some(body))
-    def withoutString(result: Result) = StringResult(result, None)
+  case class Hash(string: String)
+  case class RevalidatableResult(result: Result, hash: Hash)
+  object RevalidatableResult {
+    def apply(result: Result, body: String) = {
+      // hashing function from Arrays.java
+      val hashLong: Long = body.getBytes("UTF-8").foldLeft(z = 1L){
+        case (accu, nextByte) => 31 * accu + nextByte
+      }
+      new RevalidatableResult(result, Hash(hashLong.toString))
+    }
   }
 
-  def withEtag(page: Page)(stringResult: StringResult): Result = {
+  def withRevalidation(page: Page)(stringResult: RevalidatableResult)(implicit request: RequestHeader): Result = {
     val cacheSeconds = page.metadata.cacheTime.cacheSeconds
-    if (cacheableStatusCodes.contains(stringResult.result.header.status)) cacheHeaders(cacheSeconds, stringResult.result, stringResult.body) else stringResult.result
+    if (cacheableStatusCodes.contains(stringResult.result.header.status)) cacheHeaders(cacheSeconds, stringResult.result, Some((stringResult.hash, request.headers.get("If-None-Match")))) else stringResult.result
```

Comment:
This is a very long line. Maybe consider breaking it into a more traditional if-else format?